core/lib/dal/src/models/proto/mod.proto

```diff
@@ -23,3 +16,75 @@ message Payload {
   repeated Transaction transactions = 8;
   optional bool last_in_batch = 10; // required
 }
+
+message Transaction {
+  oneof common_data {
+    L1TxCommonData L1 = 1;
+    L2TxCommonData L2 = 2;
+    ProtocolUpgradeTxCommonData ProtocolUpgrade = 3;
+  }
+  optional Execute execute = 4; // required
+  optional uint64 received_timestamp_ms = 5; // required
+  optional bytes raw_bytes = 6; // optional
+}
+
+message L1TxCommonData {
+  optional bytes sender_address = 1; // required; H160
+  optional uint64 serial_id = 2; // required
+  optional uint64 deadline_block = 3; // required
+  optional bytes layer_2_tip_fee = 4; // required; U256
+  optional bytes full_fee = 5; // required; U256
+  optional bytes max_fee_per_gas = 6; // required; U256
+  optional bytes gas_limit = 7; // required; U256
+  optional bytes gas_per_pubdata_limit = 8; // required; U256
+  optional uint32 op_processing_type = 9; // required
+  optional uint32 priority_queue_type = 10; // required; U256
+  optional bytes eth_hash = 11; // required; H256
+  optional uint64 eth_block = 12; // required
+  optional bytes canonical_tx_hash = 13; // // required; H256
+  optional bytes to_mint = 14; // required; U256
+  optional bytes refund_recipient_address = 15; // required; H160
+}
+
+message L2TxCommonData {
+  optional uint32 nonce = 1; // required
+  optional bytes gas_limit = 2; // required; U256
+  optional bytes max_fee_per_gas = 3; // required; U256
+  optional bytes max_priority_fee_per_gas = 4; // required; U256
+  optional bytes gas_per_pubdata_limit = 5; // required; U256
+  optional bytes initiator_address = 6; // required; H160
+  optional bytes signature = 7; // required
+  optional uint32 transaction_type = 8; // required
+  optional InputData input = 9; // optional
+  optional PaymasterParams paymaster_params = 10; // required
+}
+
+message ProtocolUpgradeTxCommonData {
+  optional bytes sender_address = 1; // required; H160
+  optional uint32 upgrade_id = 2; // required
+  optional bytes max_fee_per_gas = 3; // required; U256
+  optional bytes gas_limit = 4; // required; U256
+  optional bytes gas_per_pubdata_limit = 5; // required; U256
+  optional bytes eth_hash = 6; // required; U256
+  optional uint64 eth_block = 7; // required
+  optional bytes canonical_tx_hash = 8; // required; H256
+  optional bytes to_mint = 9; // required; U256
+  optional bytes refund_recipient_address = 10; // required; H160
+}
+
+message Execute {
+  optional bytes contract_address = 1; // required; H160
+  optional bytes calldata = 2; // required
+  optional bytes value = 3; // required; U256
+  repeated bytes factory_deps = 4; // optional
+}
+
+message InputData {
+  optional bytes data = 1; // required;
+  optional bytes hash = 2; // required; H256
+}
```

Comment:
> ...but it's somewhat justified given that we need to be able to query transaction fields

That justifies the storing of the normalized data, but what about the storing the raw data? I guess it's a tradeoff to re-encoding. But is that currently used?

Even if not, I'm not sure it makes sense to remove it from the protobuf format, without also removing it entirely from `zksync_types::Transaction` and the DB table, because it will create discrepancies during the lifecycle of the object. Same with the top-level `raw_bytes` which is optional because it's only relevant for L2 transactions, while I'm not sure if it's indeed required in that case. `input` _is_ required due to the hash field (and maybe also due to some other usage), but it can be fixed. 

Another option is to create a derivative struct for `zksync_types::Transaction` which won't have these fields, and will be associated with the protobuf format. But it seems like another workaround. 

Any thoughts? @slowli @pompon0

Reply:
> but what about the storing the raw data

I may be mistaken, but I think that fields -> raw data conversion is not always possible, or more precisely, it isn't guaranteed to arrive at identical raw data (i.e., raw data -> fields conversion may be lossy). This would justify Postgres, and to lesser degree, in-memory transaction representation. I may be completely wrong about this though; cc @RomanBrodetski @popzxc

Reply:
Raw data <-> [`TransactionRequest`](https://github.com/matter-labs/zksync-era/blob/98f87cddaf4a41591d5f9861d52d3a21e698e9c4/core/lib/types/src/transaction_request.rs#L202) are expected to *not* be lossy, e.g. the roundtrip should be possible.
`TransactionRequest` -> `L2Tx` is lossy, e.g. it misses chain ID (there actually exists a method to convert `L2Tx` to RLP: [`L2Tx::get_rlp_bytes`](https://github.com/matter-labs/zksync-era/blob/98f87cddaf4a41591d5f9861d52d3a21e698e9c4/core/lib/types/src/l2/mod.rs#L237) which only takes said chain ID as an argument, so this may be the only difference, but I didn't check that.

> I don't know how scary is zksync_types::Transaction refactoring

Very scary, and comes with a high number of wtf/minute. Though if some brave soul would commit to that, I think it may make `zksync_types` much better.

Reply:
> As far as I'm concerned we can have yet another type for representing a normalized transaction, if that's useful.

We can leave the existing codebase as is for the moment, but if we care about normalization, we can implement it for the `Transaction` message, right? @moshababo I'm fine with you doing it in a follow-up PR (although IIUC, it'll mean that some work in this PR – such as structured L2 tx representation in Proobuf – is obsolete), just want to make sure that it'll be done eventually.